network/msgCompressor.go

```diff
@@ -0,0 +1,151 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package network
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	"github.com/DataDog/zstd"
+
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+var zstdCompressionMagic = [4]byte{0x28, 0xb5, 0x2f, 0xfd}
+
+// checkCanCompress checks if there is an proposal payload message and peers supporting compression
+func checkCanCompress(request broadcastRequest, prio bool, peers []*wsPeer) bool {
+	canCompress := false
+	if prio {
+		hasPP := false
+		for _, tag := range request.tags {
+			if tag == protocol.ProposalPayloadTag {
+				hasPP = true
+				break
+			}
+		}
+		// if have proposal payload check if there are any peers supporting compression
+		if hasPP {
+			for _, peer := range peers {
+				if peer.vfCompressedProposalSupported() {
+					canCompress = true
+					break
+				}
+			}
+		}
+	}
+	return canCompress
+}
+
+// zstdCompressMsg returns a concatenation of a tag and compressed data
+func zstdCompressMsg(tbytes []byte, d []byte) ([]byte, string) {
+	bound := zstd.CompressBound(len(d))
+	mbytesComp := make([]byte, len(tbytes)+bound)
+	copy(mbytesComp, tbytes)
+	comp, err := zstd.Compress(mbytesComp[len(tbytes):], d)
+	if err != nil {
+		// fallback and reuse non-compressed original data
+		logMsg := fmt.Sprintf("failed to compress into buffer of len %d: %v", len(d), err)
+		copied := copy(mbytesComp[len(tbytes):], d)
+		return mbytesComp[:len(tbytes)+copied], logMsg
```

Comment:
I think you're trying to return the uncompressed data here, but the destination buffer might be small (if the bound given by `CompressBound` is smaller than `len(d)`). Then the copy would stop early.

(But looking at the implementation, I guess CompressBound is always _bigger_.)

This place also looks like a candidate for sync.Pools in a later PR.

Reply:
this is correct, CompressBound is bugger. I added `bound` adjustment for a case if CompressBound implementation changes.